Main.elm

```diff
@@ -14,52 +15,65 @@ import Json.Encode
 type Msg
     = NewEmail String
     | NewPassphrase String
-    | NewData String
+    | NewData (Maybe String)
+    | UpdateContent String
     | NewError String
     | GetData
     | Lock
-    | SetData String
     | DataSaved String
     | DataNotSaved String
-    | UnStored Time.Time
+    | Modifying Time.Time
     | BlurSelection
     | CopySelection
     | ToggleReveal
+    | LastModified Time.Time
 
 
 type alias Model =
     { lock : Bool
     , email : String
     , passphrase : String
     , content : String
-    , stored : Bool
+    , loadedContent :
+        -- may be desynchronized with "content", only used to redraw the
+        -- contentEditable with new decrypted content
+        String
+    , modified : Bool
     , error : String
     , reveal : Bool
+    , lastModified : Maybe Time.Time
     }
 
 
 init : ( Model, Cmd msg )
 init =
-    Model True "" "" "" False "" False ! []
-
-
-
--- Handle contenteditable events
-
-
-innerHtmlDecoder =
-    Json.Decode.at [ "target", "innerHtml" ] Json.Decode.string
+    Model True "" "" "" "" False "" False Nothing ! []
 
 
 
 -- Update
 
 
-update : Msg -> Model -> ( Model, Cmd msg )
+update : Msg -> Model -> ( Model, Cmd Msg )
 update message model =
     case message of
-        UnStored time ->
-            { model | stored = False } ! []
+        Modifying now ->
+            -- The content is being edited, should we save? (aka debounce)
```

Comment:
There's a [few](http://package.elm-lang.org/packages/athanclark/elm-debouncer/latest/Debounce) [existing](http://package.elm-lang.org/packages/jinjor/elm-debounce/latest) debouncers already, but I didn't feel the extra overhead was worth it.

There's something I stumbled upon lately though, that I might want to use instead of my current solution based on subscriptions: https://github.com/fredcy/example-elm-debounce/blob/master/Debounce.elm#L48. The idea is to fire a `Process.sleep` (the equivalent of a `setTimeout`) each time a key is pressed, and only react on the very last one.

I think I'm going to use that instead, it feels very natural.